R/importance.R

```diff
@@ -36,131 +39,243 @@
+        is.null(names(covariate_groups)) | names(covariate_groups) == "" |
+          is.na(names(covariate_groups))
+      ))
+      if (any(sapply(covariate_groups[unique(no_name)], length)) != 1) {
+        stop("Covariate groups with more than one covariate must be named.")
+      } else if (all(sapply(covariate_groups[unique(no_name)], length)) == 1) {
+        names(covariate_groups[no_name]) <- unlist(covariate_groups[no_name])
+      }
+    }
+    X <- covariate_groups # for convenience in remaining parts of function
+  } else {
+    names(X) <- X
+  }
+
+  ########################## risk under observed data ##########################
+  original_pred <- fit$predict_fold(task, fold_number = fold_number)
+  original_eval <- eval_fun(original_pred, Y)
+  if (!is.null(attr(original_eval, "loss")) && !attr(original_eval, "loss")) {
+    original_risk <- original_eval
+  } else {
+    original_losses <- original_eval
+    original_risk <- mean(original_losses)
+  }
+
+  ######################## list of importance results ##########################
+  res_list <- lapply(seq_along(X), function(i) {
+    # get the relevant group/covariate
+    if (!is.null(covariate_groups)) {
+      x <- X[[i]]
+    } else {
+      x <- X[i]
+    }
+    # get the risk when the covariate/group is permuted/removed
     if (type == "permute") {
-      # permute covariate x, and give it the same name as the original x
-      x_permuted <- data.table(sample(unlist(d[, x, with = FALSE]), nrow(d)))
-      names(x_permuted) <- x
-      # replace original x with permuted x, and update task with permuted x
-      x_permuted_name <- task$add_columns(x_permuted)
-      task_x_permuted <- task$next_in_chain(column_names = x_permuted_name)
-      # obtain predictions & risk on the new task with permuted x
-      x_permuted_pred <- fit$predict_fold(task_x_permuted, fold_number)
-      no_x_risk <- mean(loss(x_permuted_pred, Y))
+      seeds <- sample(11:2^15, B)
+      no_x_risk <- mean(unlist(parallel::mclapply(seeds, function(b) {
+        set.seed(b)
+        # get the permutation rows
+        perm <- sample(1:nrow(d), nrow(d))
+        # permute x (the covariate/group), and name it as the original x
+        x_perm <- data.table(d[perm, x, with = FALSE])
+        names(x_perm) <- x
+        # replace original x with permuted x, and update task with permuted x
+        x_perm_name <- task$add_columns(x_perm)
+        task_x_perm <- task$next_in_chain(column_names = x_perm_name)
+        # obtain predictions & risk on the new task with permuted x
+        x_perm_pred <- fit$predict_fold(task_x_perm, fold_number)
+        x_perm_eval <- eval_fun(x_perm_pred, Y)
+        if (!is.null(attr(original_eval, "loss")) && !attr(original_eval, "loss")) {
+          no_x_risk <- x_perm_eval
+        } else {
+          no_x_losses <- x_perm_eval
+          no_x_risk <- mean(no_x_losses)
+        }
+        return(no_x_risk)
+      }, mc.cores = cores)))
```

Comment:
`future.apply::future_lapply` will make specification of this option moot, but you should add `future.seed = TRUE` in its place.